Services/dev

pankhuri-mt opened the pull request:
The PR brings in ServerBase component of mtrix package renamed to Service with auto generation of connection managers and stress test inclusion. I found the below summary helpful added in PR #48 and #55. The information can also be found in README for Services component.

# Auto-generation for Connection Managers

This PR implements an auto-generator for ConnectionManagers. It adds a `generate_connection_manager` method to `mindtrace.services.core.utils` that can be called by a `Service` class to auto-generate a ConnectManager, which may be returned upon calling the service's `connect` method.

The main idea is that now when calling a service's `connect` method, the following logic is followed:

```python
if cls._client_interface is None:
    return generate_connection_manager(cls)(url=url)
else:
    return cls._client_interface(url=url)
```

That is, the same `Service`/`ConnectionManager` relationship exists as before, but now if a `ConnectionManager` is not explicitly registered to a class, it is auto-generated every time the service's `connect` method is called. If a `ConnectionManager` class _is_ registered to a `Service` class, it defaults to previous behavior, returning the registered ConnectionManager class instead.

## Updated `add_endpoint` method

To aid the new `generate_connection_manager` in generating its magic, when calling `Service.add_endpoint()`, a `schema: TaskSchema` must now be passed in, which is saved to the service's `endpoints` dict (which was before a list). The generate method makes use of each endpoint's schema to add type validation as well as defining the returned ConnectionManager's methods with the right arguments. 

Note that all endpoints added via `add_endpoint` are, by default, put through the general logic of generating connection manager methods for Service endpoints, except for two: the `status` and `shutdown` methods, which require special logic for when the service is _not_ running (the methods must still work), and as such remain defined by the base ConnectionManager class. 

## Both `GET` and `POST` requests default to connection manager methods

A slight modification from previous `Service` implementations— all generated endpoints are now methods in the returned connection manager, with naked properties not currently supported. That is, 

```python
from mindtrace.services import Service
cm = Service.launch()
cm.status  # no longer supported
cm.status()  # it is now generated as a method
```

This behavior can be modified as desired in future iterations.

## Usage example

```python
from mindtrace.services import Service

cm = Service.launch()

cm.status()  # StatusOutput(status=<ServerStatus.Available: 'Available'>)
cm.heartbeat()  # HeartbeatOutput(heartbeat=Heartbeat(status=<ServerStatus.Available: 'Available'>, server_id=UUID('355143ec-5075-11f0-90c6-677c4904e0b4'), message='Heartbeat check successful.', details=None))
cm.endpoints()  # EndpointsOutput(endpoints=['endpoints', 'status', 'heartbeat', 'server_id', 'pid_file', 'shutdown'])
cm.server_id()  # ServerIDOutput(server_id=UUID('355143ec-5075-11f0-90c6-677c4904e0b4'))
cm.pid_file()  # PIDFileOutput(pid_file='~/.cache/mindtrace/pids/Service_355143ec-5075-11f0-90c6-677c4904e0b4_pid.txt')
cm.shutdown(block=True)  # ShutdownOutput(shutdown=True)
```

## Creating Custom Services

To create a custom class, subclass the `Service` class, as usual:

```python
class EchoService(Service):
    def __init__(self, *args, **kwargs):
        super().__init__(*args, **kwargs)
        self.add_endpoint("echo", self.echo, schema=echo_task)

    def echo(self, payload: EchoInput) -> EchoOutput:
        if payload.delay > 0:
            time.sleep(payload.delay)
        return EchoOutput(echoed=payload.message)
```

As mentioned, the connection manager is now autogenerated:

```python
from mindtrace.services.sample.echo_service import EchoService

cm = EchoService.launch()
cm.echo(message="Hello, world!")  # EchoOutput(echoed='Hello, world!')
```

Note there is some magic in `generate_connection_manager` that makes the fields of the input schema registered to an endpoint the kwargs of that arguments message, and which are required. I.e. the following fails:

```
cm.echo("Hello, world!")  # Error: args are not supported, use kwargs, as defined by the `echo_task` schema
```

### Input and Output Schemas

Note that in defining custom endpoints, schemas are now required. Currently, the expected schema is a `TaskSchema`, now defined in `mindtrace.core`, which is more or less identical to the `JobSchema` in previous iterations. I.e. here is the actual definition for the EchoService class:

```python
class EchoInput(BaseModel):
    message: str
    delay: float = 0.0 

class EchoOutput(BaseModel):
    echoed: str

class EchoTaskSchema(TaskSchema):
    name: str = "echo"
    input_schema: type[EchoInput] = EchoInput
    output_schema: type[EchoOutput] = EchoOutput

echo_task = EchoTaskSchema()


class EchoService(Service):
    ...
```

Notice that you must define (subclass) a `TaskSchema`; `input_schema` and `output_schema` are both optional, but if present must be derived from `BaseModel`. The input and output schemas can be used to validate input / outputs automatically (which is True by default):

```python
def generate_connection_manager():
    # Buried in a couple layers of decorator magic:
    if not validate_output:
        return result  # raw result dict
    return output_schema(**result) if output_schema is not None else result
```

### Async Support

Async support is included in the current iteration.

```python
import asyncio

from mindtrace.core import TimerCollection
from mindtrace.services.sample.echo_service import EchoService

def sync_example_with_delay(cm, timers):
    timers.start("Synchronous")
    for i in range(5):
        result = cm.echo(message=f"Sync message {i+1}", delay=1.0)
    timers.stop("Synchronous")
        
async def async_example_with_delay(cm, timers):
    tasks = [cm.aecho(message=f"Async message {i+1}", delay=1.0) for i in range(5)]
    timers.start("Asynchronous")
    results = await asyncio.gather(*tasks)    
    timers.stop("Asynchronous")

def main():
    timers = TimerCollection()

    with EchoService.launch(timeout=15) as cm:  # note num_workers=1 is fine here, because gunicorn will thread for us
        # Run sync example (should take ~5 seconds)
        sync_example_with_delay(cm, timers)
        
        # Run async example (should take ~1 second)
        asyncio.run(async_example_with_delay(cm, timers))
    
    print("\n" + "=" * 50)
    print("SUMMARY")
    print("=" * 50)

    print(timers)

if __name__ == "__main__":
    main()
```

```text
==================================================
SUMMARY
==================================================
Synchronous: 5.193916s
Asynchronous: 1.094834s
```

## Documentation & Testing

```bash
The behavior of `ds test` is now:
- `ds test` by itself runs all test suites: `unit`, `integration`, and `stress` (identical to before, but now there is an added `stress` suite)

- `ds test` + any number of individual flags will run just those suites. E.g.
- - `ds test --unit --stress` will run both the unit and stress test suites.

Test suites are run individually, but coverage is appended for later suites. The stress test suite provides verbose output (i.e. tqdm progress bars), as it's quite lengthy. Also, currently the stress test suite has difficulty stopping early if it hits an error.

Some of the tests get quite long, here are my current times for each suite:

```text
`unit`: 522 passed, 5 skipped in 4.56s
`unit`: 527 passed in 9.69s  # w/ `uv add torch datasets transformers`
`integration`: 58 passed in 41.78s
`stress`: 7 passed in 208.89s (0:03:28)
```

All test suites should pass, with `ds test --unit` yielding ~97% coverage, and `ds test --unit --integration` yielding 100%.

Commit:
Add generate_connection_manager method to autogen connection managers for services

Commit:
Update the generate_connection_manager method to work with async methods

Commit:
Add Service related unit and integration tests

pankhuri-mt committed:
initial commit

Commit:
Copy Service class over from latest mtrix release

Commit:
Fix logger errors

Commit:
Add a unit test suite for the helper methods in mindtrace.core.utils.timers

Commit:
Add a test suite for the mindtrace.core.utils.lambdas module

Commit:
Add additional unit test for ifnone

Commit:
Add TaskSchema class

Commit:
Fix mis-typed TypeVar

Commit:
Add a working generate_connection_manager method to auto-generate connection managers

Commit:
Remove extraneous commented line

Commit:
Add pydantic as a core dependency

pankhuri-mt committed:
Merge pull request #44 from Mindtrace/services/feature/cm-generator

Initial copy of Service class into mindtrace package

Commit:
Merge dev into cm-generator branch

Commit:
Merge dev into services/feature/additional-unit-tests

Commit:
Fix Service status issues

Commit:
Add two Service related samples

Commit:
Update integration test to use separate EchoService fixture

Commit:
Delete unused test file

Commit:
Rename erroneous 'blocking' argument to 'validate_output', matching its actual meaning

Commit:
Add validate_input argument on auto-generated connection manager methods

Commit:
Add unit tests associated with validate_input services argument

Commit:
Add additional unit tests for mindtrace.core.base.mindtrace_base

Commit:
Add name validation tests for logger

Commit:
Rename TestService to SampleService to avoid confusing pytest

Commit:
Rename mindtrace.services.base to mindtrace.services.core

Commit:
Add unit test suite for ConnectionManager class

Commit:
Add a services launcher unit test suite

Commit:
Add integration test for all default Service endpoints

Commit:
Add unit test suite for services types

Commit:
Add initial unit test suite for services utils module

Commit:
Add additional unit tests for the services utils module

Commit:
Add unit test suite for the EchoService class

pankhuri-mt committed:
adding asyncio in dev dependency

Commit:
Add additional unit tests for the Service class

Commit:
Add additional unit tests for the Service class

pankhuri-mt committed:
expand user logic

pankhuri-mt committed:
add expanduser logic in Config

pankhuri-mt committed:
Merge pull request #52 from Mindtrace/services/feature/additional-unit-tests

Merge dev into services/dev and add additional tests

Commit:
Update EchoService echo method to take an optional  argument

Commit:
Update EchoService unit tests to match the new delay argument

Commit:
Add additional Service unit tests

Commit:
Add initial stress test test suite

Commit:
Save stress test results to file

Commit:
Fix code coverage bug where stress tests were not combined with unit/integration test coverage

Commit:
Merge dev into services/dev

Commit:
Remove merge duplicate

pankhuri-mt committed:
Merge pull request #55 from Mindtrace/services/feature/even-more-unit-tests

Additional unit tests, notably for the services module

pankhuri-mt committed:
Merge remote-tracking branch 'origin/dev' into services/dev

pankhuri-mt committed:
readme for services

vik-rant commented on mindtrace/services/mindtrace/services/README.md:
```diff
@@ -0,0 +1,144 @@
+
+## Usage Example
+
+See [`sample/echo_service.py`](./sample/echo_service.py) for a full example. Basic usage:
```
Would we eventually want to move the sample to `/samples/services/`

vik-rant commented on scripts/run_tests.sh:
```diff
@@ -31,30 +40,116 @@ while [[ $# -gt 0 ]]; do
+# Start MinIO container if running integration tests or all tests
+if [ "$RUN_INTEGRATION" = true ]; then
+    echo "Starting docker containers..."
+    docker-compose -f tests/docker-compose.yml up -d
```
Let us use the `$DOCKER_COMPOSE_CMD` deduced above, instead of `docker-compose`.

Commit:
Remove unused job-related functionality and tests

- Remove get_job/aget_job methods from connection manager generation
- Remove all job-related unit tests (6 test methods)
- Remove job method assertions from integration tests
- Job functionality was unused (no server-side implementation exists)

pankhuri-mt committed:
remove uv lock

pankhuri-mt committed:
use the DOCKER_COMPOSE_CMD
